File: spring-security/src/main/java/com/example/spring_security/service/RoommateService.java
package com.example.spring_security.service;

import com.example.spring_security.dao.RoommateAnnouncementRepository;
import com.example.spring_security.dao.RoommateApplicationRepository;
import com.example.spring_security.dao.UserRepo;
import com.example.spring_security.dao.PropertyListingRepository;
import com.example.spring_security.dao.InquiryRepository;
import com.example.spring_security.dto.RoommateAnnouncementCreateDTO;
import com.example.spring_security.dto.RoommateAnnouncementDTO;
import com.example.spring_security.dto.RoommateApplicationDTO;
import com.example.spring_security.dto.RoommateApplicationCreateDTO;
import com.example.spring_security.dto.RoommateApplicationResponseDTO;
import com.example.spring_security.dto.UserBasicDTO;import com.example.spring_security.dto.InquiryDTO;import com.example.spring_security.dto.PropertyListingDTO;import com.example.spring_security.dto.PropertyListingBasicDTO;
import com.example.spring_security.exception.AccessDeniedException;
import com.example.spring_security.exception.ResourceNotFoundException;
import com.example.spring_security.model.RoommateAnnouncement;
import com.example.spring_security.model.RoommateApplication;
import com.example.spring_security.model.User;
import com.example.spring_security.model.UserPrincipal;
import com.example.spring_security.model.PropertyListing;
import com.example.spring_security.model.Inquiry;
import com.example.spring_security.model.InquiryStatus;
import com.example.spring_security.service.CompatibilityService;
import com.example.spring_security.service.EmailService;
import com.example.spring_security.service.WebSocketNotificationService;
import org.modelmapper.ModelMapper;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.time.LocalDateTime;

/**
 * Service for managing roommate announcements and applications
 * Includes ML-powered compatibility scoring and recommendations
 */
@Service
@Transactional(readOnly = true)
public class RoommateService {
    
    private static final Logger logger = LoggerFactory.getLogger(RoommateService.class);
    
    private final RoommateAnnouncementRepository announcementRepository;
    private final RoommateApplicationRepository applicationRepository;
    private final UserRepo userRepository;
    private final PropertyListingRepository propertyListingRepository;
    private final InquiryRepository inquiryRepository;
    private final EmailService emailService;
    private final ModelMapper modelMapper;
    private final CompatibilityService compatibilityService;
    private final WebSocketNotificationService webSocketNotificationService;
    
    @Autowired
    public RoommateService(RoommateAnnouncementRepository announcementRepository,
                          RoommateApplicationRepository applicationRepository,
                          UserRepo userRepository,
                          PropertyListingRepository propertyListingRepository,
                          InquiryRepository inquiryRepository,
                          EmailService emailService,
                          ModelMapper modelMapper,
                          CompatibilityService compatibilityService,
                          WebSocketNotificationService webSocketNotificationService) {
        this.announcementRepository = announcementRepository;
        this.applicationRepository = applicationRepository;
        this.userRepository = userRepository;
        this.propertyListingRepository = propertyListingRepository;
        this.inquiryRepository = inquiryRepository;
        this.emailService = emailService;
        this.modelMapper = modelMapper;
        this.compatibilityService = compatibilityService;
        this.webSocketNotificationService = webSocketNotificationService;
    }
    
    /**
     * Create a new roommate announcement
     */
    @Transactional
    public RoommateAnnouncementDTO createAnnouncement(RoommateAnnouncementCreateDTO createDTO, UserPrincipal currentUser) {
        logger.info("Creating roommate announcement by user {}", currentUser.getId());
        
        User poster = userRepository.findById(currentUser.getId())
                .orElseThrow(() -> new ResourceNotFoundException("User not found"));
        
        if (poster.getRole() != User.Role.STUDENT) {
            throw new AccessDeniedException("Only students can post roommate announcements");
        }
        
        RoommateAnnouncement announcement = RoommateAnnouncement.builder()
                .poster(poster)
                // Property details
                .propertyTitle(createDTO.getPropertyTitle())
                .propertyAddress(createDTO.getPropertyAddress())
                .propertyLatitude(createDTO.getPropertyLatitude())
                .propertyLongitude(createDTO.getPropertyLongitude())
                .totalRent(createDTO.getTotalRent())
                .totalRooms(createDTO.getTotalRooms())
                .availableRooms(createDTO.getAvailableRooms())
                .propertyType(createDTO.getPropertyType() != null ? 
                    RoommateAnnouncement.PropertyType.valueOf(createDTO.getPropertyType()) : 
                    RoommateAnnouncement.PropertyType.APARTMENT)
                // Roommate preferences
                .maxRoommates(createDTO.getMaxRoommates())
                .genderPreference(createDTO.getGenderPreference() != null ? 
                    RoommateAnnouncement.GenderPreference.valueOf(createDTO.getGenderPreference()) : 
                    RoommateAnnouncement.GenderPreference.NO_PREFERENCE)
                .ageMin(createDTO.getAgeMin() != null ? createDTO.getAgeMin() : 18)
                .ageMax(createDTO.getAgeMax() != null ? createDTO.getAgeMax() : 35)
                .lifestyleTags(createDTO.getLifestyleTags() != null ? createDTO.getLifestyleTags() : new java.util.HashSet<>())
                .smokingAllowed(createDTO.getSmokingAllowed() != null ? createDTO.getSmokingAllowed() : false)
                .petsAllowed(createDTO.getPetsAllowed() != null ? createDTO.getPetsAllowed() : false)
                .cleanlinessLevel(createDTO.getCleanlinessLevel() != null ? createDTO.getCleanlinessLevel() : 3)
                // Financial details
                .rentPerPerson(createDTO.getRentPerPerson())
                .securityDeposit(createDTO.getSecurityDeposit() != null ? createDTO.getSecurityDeposit() : BigDecimal.ZERO)
                .utilitiesSplit(createDTO.getUtilitiesSplit() != null ? 
                    RoommateAnnouncement.UtilitiesSplit.valueOf(createDTO.getUtilitiesSplit()) : 
                    RoommateAnnouncement.UtilitiesSplit.EQUAL)
                .additionalCosts(createDTO.getAdditionalCosts())
                // Posting details
                .description(createDTO.getDescription())
                .moveInDate(createDTO.getMoveInDate())
                .leaseDurationMonths(createDTO.getLeaseDurationMonths())
                .status(RoommateAnnouncement.AnnouncementStatus.ACTIVE)
                .expiresAt(LocalDateTime.now().plusDays(90))
                .build();
        
        announcement = announcementRepository.save(announcement);
        return convertAnnouncementToDTO(announcement, currentUser.getId());
    }
    
    /**
     * Get announcements for browsing
     */
    public Page<RoommateAnnouncementDTO> getAnnouncementsForBrowsing(UserPrincipal currentUser, Pageable pageable) {
        logger.info("Fetching announcements for browsing by user {}", currentUser.getId());
        
        Page<RoommateAnnouncement> announcements = announcementRepository.findByPosterIdNotAndStatusAndExpiresAtAfter(
                currentUser.getId(),
                RoommateAnnouncement.AnnouncementStatus.ACTIVE,
                LocalDateTime.now(),
                pageable);
        
        return announcements.map(announcement -> convertAnnouncementToDTO(announcement, currentUser.getId()));
    }
    
    /**
     * Get my announcements
     */
    public Page<RoommateAnnouncementDTO> getMyAnnouncements(UserPrincipal currentUser, Pageable pageable) {
        Page<RoommateAnnouncement> announcements = announcementRepository.findByPosterIdOrderByCreatedAtDesc(
                currentUser.getId(), pageable);
        
        return announcements.map(announcement -> convertAnnouncementToDTO(announcement, currentUser.getId()));
    }
    
    /**
     * Get announcement by ID
     */
    public RoommateAnnouncementDTO getAnnouncementById(Long announcementId, UserPrincipal currentUser) {
        RoommateAnnouncement announcement = announcementRepository.findByIdWithDetails(announcementId)
                .orElseThrow(() -> new ResourceNotFoundException("Announcement not found"));
        
        return convertAnnouncementToDTO(announcement, currentUser.getId());
    }
    
    /**
     * Apply to announcement with ML compatibility scoring
     */
    @Transactional
    public RoommateApplicationDTO applyToAnnouncement(RoommateApplicationCreateDTO createDTO, UserPrincipal currentUser) {
        User applicant = userRepository.findById(currentUser.getId())
                .orElseThrow(() -> new ResourceNotFoundException("User not found"));
        
        RoommateAnnouncement announcement = announcementRepository.findByIdWithDetails(createDTO.getAnnouncementId())
                .orElseThrow(() -> new ResourceNotFoundException("Announcement not found"));
        
        // Basic validation
        if (announcement.getPoster().getId() == currentUser.getId()) {
            throw new IllegalStateException("You cannot apply to your own announcement");
        }
        
        // Calculate ML compatibility score
        BigDecimal compatibilityScore = null;
        try {
            compatibilityScore = compatibilityService.calculateCompatibility(applicant, announcement);
            logger.info("Calculated compatibility score {} for user {} applying to announcement {}", 
                       compatibilityScore, applicant.getId(), announcement.getId());
        } catch (Exception e) {
            logger.warn("Failed to calculate compatibility score for application: {}", e.getMessage());
            compatibilityScore = BigDecimal.valueOf(0.50); // Default neutral score
        }
        
        RoommateApplication application = RoommateApplication.builder()
                .announcement(announcement)
                .applicant(applicant)
                .poster(announcement.getPoster())
                .message(createDTO.getMessage())
                .compatibilityScore(compatibilityScore)
                .status(RoommateApplication.ApplicationStatus.PENDING)
                .build();
        
        application = applicationRepository.save(application);
        
        // Send notification to poster with compatibility score
        sendApplicationNotificationWithScore(application, compatibilityScore);
        
        return convertApplicationToDTO(application);
    }
    
    /**
     * Get applications for announcement
     */
    public Page<RoommateApplicationDTO> getApplicationsForAnnouncement(Long announcementId, UserPrincipal currentUser, Pageable pageable) {
        Page<RoommateApplication> applications = applicationRepository.findByAnnouncementIdOrderByAppliedAtDesc(
                announcementId, pageable);
        
        return applications.map(this::convertApplicationToDTO);
    }
    
    /**
     * Get my applications
     */
    public Page<RoommateApplicationDTO> getMyApplications(UserPrincipal currentUser, Pageable pageable) {
        Page<RoommateApplication> applications = applicationRepository.findByApplicantIdOrderByAppliedAtDesc(
                currentUser.getId(), pageable);
        
        return applications.map(this::convertApplicationToDTO);
    }
    
    /**
     * Respond to application
     */
    @Transactional
    public RoommateApplicationDTO respondToApplication(Long applicationId, RoommateApplicationResponseDTO responseDTO, UserPrincipal currentUser) {
        RoommateApplication application = applicationRepository.findByIdWithDetails(applicationId)
                .orElseThrow(() -> new ResourceNotFoundException("Application not found"));
        
        if (application.getPoster().getId() != currentUser.getId()) {
            throw new AccessDeniedException("You can only respond to applications for your own announcements");
        }
        
        RoommateApplication.ApplicationStatus newStatus = RoommateApplication.ApplicationStatus.valueOf(responseDTO.getStatus().toUpperCase());
        
        if (newStatus == RoommateApplication.ApplicationStatus.ACCEPTED) {
            application.accept(responseDTO.getResponseMessage());
        } else if (newStatus == RoommateApplication.ApplicationStatus.REJECTED) {
            application.reject(responseDTO.getResponseMessage());
        }
        
        application = applicationRepository.save(application);
        
        // Send WebSocket notification to applicant about response
        boolean accepted = newStatus == RoommateApplication.ApplicationStatus.ACCEPTED;
        webSocketNotificationService.notifyOfRoommateApplicationResponse(
            application.getApplicant().getId(),
            convertApplicationToDTO(application),
            accepted
        );
        
        return convertApplicationToDTO(application);
    }
    
    /**
     * Get closed deals for student
     */
    public Page<InquiryDTO> getClosedDealsForStudent(UserPrincipal currentUser, Pageable pageable) {
        Page<Inquiry> closedDeals = inquiryRepository.findByStudentIdAndStatusOrderByReplyTimestampDesc(
                Long.valueOf(currentUser.getId()), InquiryStatus.CLOSED, pageable);
        
        return closedDeals.map(this::convertInquiryToClosedDealDTO);
    }
    
    // Helper conversion methods
    private RoommateAnnouncementDTO convertAnnouncementToDTO(RoommateAnnouncement announcement, Integer currentUserId) {
        return RoommateAnnouncementDTO.builder()
                .id(announcement.getId())
                .poster(convertUserToBasicDTO(announcement.getPoster()))
                // Property details
                .propertyTitle(announcement.getPropertyTitle())
                .propertyAddress(announcement.getPropertyAddress())
                .totalRent(announcement.getTotalRent())
                .totalRooms(announcement.getTotalRooms())
                .availableRooms(announcement.getAvailableRooms())
                .propertyType(announcement.getPropertyType() != null ? announcement.getPropertyType().name() : null)
                // Roommate preferences
                .maxRoommates(announcement.getMaxRoommates())
                .genderPreference(announcement.getGenderPreference() != null ? announcement.getGenderPreference().name() : null)
                .ageMin(announcement.getAgeMin())
                .ageMax(announcement.getAgeMax())
                .lifestyleTags(announcement.getLifestyleTags())
                .smokingAllowed(announcement.getSmokingAllowed())
                .petsAllowed(announcement.getPetsAllowed())
                .cleanlinessLevel(announcement.getCleanlinessLevel())
                // Financial details
                .rentPerPerson(announcement.getRentPerPerson())
                .securityDeposit(announcement.getSecurityDeposit())
                .utilitiesSplit(announcement.getUtilitiesSplit() != null ? announcement.getUtilitiesSplit().name() : null)
                .additionalCosts(announcement.getAdditionalCosts())
                // Posting details
                .description(announcement.getDescription())
                .moveInDate(announcement.getMoveInDate())
                .leaseDurationMonths(announcement.getLeaseDurationMonths())
                .status(announcement.getStatus().name())
                .createdAt(announcement.getCreatedAt())
                .expiresAt(announcement.getExpiresAt())
                // Helper fields
                .remainingSpots(announcement.getRemainingSpots())
                .applicationCount(announcement.getApplications() != null ? announcement.getApplications().size() : 0)
                .build();
    }
    
    private RoommateApplicationDTO convertApplicationToDTO(RoommateApplication application) {
        return RoommateApplicationDTO.builder()
                .id(application.getId())
                .announcementId(application.getAnnouncement().getId())
                .announcementTitle(application.getAnnouncement().getPropertyTitle())
                .applicant(convertUserToBasicDTO(application.getApplicant()))
                .poster(convertUserToBasicDTO(application.getPoster()))
                .message(application.getMessage())
                .compatibilityScore(application.getCompatibilityScore())
                .status(application.getStatus().name())
                .appliedAt(application.getAppliedAt())
                .respondedAt(application.getRespondedAt())
                .responseMessage(application.getResponseMessage())
                .build();
    }
    
    private UserBasicDTO convertUserToBasicDTO(User user) {
        return UserBasicDTO.builder()
                .id(user.getId())
                .username(user.getUsername())
                .email(user.getEmail())
                .role(user.getRole().name())
                .build();
    }
    
    private InquiryDTO convertInquiryToBasicDTO(Inquiry inquiry) {
        return InquiryDTO.builder()
                .id(inquiry.getId())
                .student(convertUserToBasicDTO(inquiry.getStudent()))
                .owner(convertUserToBasicDTO(inquiry.getOwner()))
                .message(inquiry.getMessage())
                .timestamp(inquiry.getTimestamp())
                .status(inquiry.getStatus().name())
                .build();
    }
    
    private InquiryDTO convertInquiryToClosedDealDTO(Inquiry inquiry) {
        PropertyListing property = inquiry.getProperty();
        
        return InquiryDTO.builder()
                .id(inquiry.getId())
                .student(convertUserToBasicDTO(inquiry.getStudent()))
                .owner(convertUserToBasicDTO(inquiry.getOwner()))
                .property(PropertyListingBasicDTO.builder()
                        .id(property.getId())
                        .title(property.getTitle())
                        .price(property.getPrice())
                        .location(property.getLocation())
                        .city(property.getCity())
                        .propertyType(property.getPropertyType())
                        .bedrooms(property.getBedrooms())
                        .bathrooms(property.getBathrooms())
                        .imageUrls(property.getImageUrls())
                        .build())
                .message(inquiry.getMessage())
                .timestamp(inquiry.getTimestamp())
                .replyTimestamp(inquiry.getReplyTimestamp())
                .status(inquiry.getStatus().name())
                .agreedPrice(property.getPrice()) // Use property price as agreed price
                .build();
    }
    
    /**
     * Send enhanced application notification with ML compatibility score
     */
    private void sendApplicationNotificationWithScore(RoommateApplication application, BigDecimal compatibilityScore) {
        try {
            String compatibilityLevel = getCompatibilityLevel(compatibilityScore);
            
            String subject = String.format("New Roommate Application - %s Compatibility Match", compatibilityLevel);
            
            String emailBody = String.format(
                "You have received a new roommate application for your announcement: %s\n\n" +
                "Applicant: %s\n" +
                "Compatibility Score: %d%% (%s)\n" +
                "Message: %s\n\n" +
                "View and respond to this application in your dashboard.",
                application.getAnnouncement().getPropertyTitle(),
                application.getApplicant().getUsername(),
                compatibilityScore.multiply(BigDecimal.valueOf(100)).intValue(),
                compatibilityLevel,
                application.getMessage()
            );
            
            // Send email notification (if emailService is properly implemented)
            // emailService.sendNotificationEmail(application.getPoster().getEmail(), subject, emailBody);
            
            // Send real-time WebSocket notification with ML compatibility score
            webSocketNotificationService.notifyOfNewRoommateApplication(
                application.getPoster().getId(),
                convertApplicationToDTO(application),
                compatibilityScore.doubleValue()
            );
            
            logger.info("Sent application notification with compatibility score to user {}", 
                       application.getPoster().getId());
            
        } catch (Exception e) {
            logger.error("Failed to send application notification: {}", e.getMessage());
        }
    }
    
    /**
     * Get compatibility level description from score
     */
    private String getCompatibilityLevel(BigDecimal score) {
        if (score == null) return "Unknown";
        
        if (score.compareTo(BigDecimal.valueOf(0.90)) >= 0) return "Excellent";
        if (score.compareTo(BigDecimal.valueOf(0.75)) >= 0) return "Very Good";
        if (score.compareTo(BigDecimal.valueOf(0.60)) >= 0) return "Good";
        if (score.compareTo(BigDecimal.valueOf(0.40)) >= 0) return "Fair";
        return "Poor";
    }
} 

File: spring-security/frontend/ui/src/app/services/roommate.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpHeaders, HttpParams } from '@angular/common/http';
import { Observable, throwError } from 'rxjs';
import { catchError, tap } from 'rxjs/operators';
import { AuthService } from '../auth/auth.service';

// Temporary interfaces - will be moved to separate model files later
export interface RoommateAnnouncementCreateDTO {
  // Optional - for Type A announcements (based on closed deals)
  propertyListingId?: number;
  
  // Property Details (required for Type B, auto-filled for Type A)
  propertyTitle: string;
  propertyAddress: string;
  propertyLatitude?: number;
  propertyLongitude?: number;
  totalRent: number;
  totalRooms: number;
  availableRooms: number;
  propertyType: string;
  
  // Roommate Preferences
  maxRoommates: number;
  genderPreference: string;
  ageMin: number;
  ageMax: number;
  lifestyleTags: string[];
  smokingAllowed: boolean;
  petsAllowed: boolean;
  cleanlinessLevel: number;
  
  // Financial Details
  rentPerPerson: number;
  securityDeposit: number;
  utilitiesSplit: string;
  additionalCosts?: string;
  
  // Posting Details
  description: string;
  moveInDate: string; // ISO date string
  leaseDurationMonths: number;
}

export interface RoommateAnnouncementDTO {
  id: number;
  poster: {
    id: number;
    username: string;
    email: string;
    role: string;
  };
  propertyTitle: string;
  propertyAddress: string;
  totalRent: number;
  totalRooms: number;
  availableRooms: number;
  maxRoommates: number;
  rentPerPerson: number;
  securityDeposit: number;
  description: string;
  moveInDate: string;
  leaseDurationMonths: number;
  status: string;
  createdAt: string;
  expiresAt: string;
  remainingSpots: number;
  applicationCount: number;
}

export interface RoommateApplicationCreateDTO {
  announcementId: number;
  message: string;
}

export interface RoommateApplicationDTO {
  id: number;
  applicant: {
    id: number;
    username: string;
    email: string;
    role: string;
  };
  poster: {
    id: number;
    username: string;
    email: string;
    role: string;
  };
  message: string;
  compatibilityScore: number; // ML-calculated score (0.0 to 1.0)
  status: string;
  appliedAt: string;
  respondedAt?: string;
  responseMessage?: string;
}

export interface RoommateApplicationResponseDTO {
  status: 'ACCEPTED' | 'REJECTED';
  responseMessage: string;
}

export interface Page<T> {
  content: T[];
  pageable: {
    sort: {
      empty: boolean;
      sorted: boolean;
      unsorted: boolean;
    };
    offset: number;
    pageSize: number;
    pageNumber: number;
    paged: boolean;
    unpaged: boolean;
  };
  last: boolean;
  totalPages: number;
  totalElements: number;
  size: number;
  number: number;
  sort: {
    empty: boolean;
    sorted: boolean;
    unsorted: boolean;
  };
  first: boolean;
  numberOfElements: number;
  empty: boolean;
}

@Injectable({
  providedIn: 'root'
})
export class RoommateService {
  private apiUrl = 'http://localhost:8080/api/v1/roommates';

  constructor(
    private http: HttpClient,
    private authService: AuthService
  ) {}

  private getAuthHeaders(): HttpHeaders {
    const token = this.authService.getToken();
    if (!token) {
      console.error('Authentication token not found.');
      return new HttpHeaders();
    }
    return new HttpHeaders().set('Authorization', `Bearer ${token}`);
  }

  // ========== ANNOUNCEMENT OPERATIONS ==========

  /**
   * Create a new roommate announcement
   * POST /api/v1/roommates/announcements
   */
  createAnnouncement(announcement: RoommateAnnouncementCreateDTO): Observable<RoommateAnnouncementDTO> {
    const headers = this.getAuthHeaders();
    return this.http.post<RoommateAnnouncementDTO>(`${this.apiUrl}/announcements`, announcement, { headers })
      .pipe(
        tap(result => console.log('Created roommate announcement:', result)),
        catchError(this.handleError)
      );
  }

  /**
   * Get announcements for browsing (excludes own announcements)
   * GET /api/v1/roommates/announcements
   */
  getAnnouncementsForBrowsing(page = 0, size = 10): Observable<Page<RoommateAnnouncementDTO>> {
    const headers = this.getAuthHeaders();
    const params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString());
    
    return this.http.get<Page<RoommateAnnouncementDTO>>(`${this.apiUrl}/announcements`, { headers, params })
      .pipe(
        tap(result => console.log(`Loaded ${result.content.length} announcements for browsing`)),
        catchError(this.handleError)
      );
  }

  /**
   * Get my announcements
   * GET /api/v1/roommates/announcements/my
   */
  getMyAnnouncements(page = 0, size = 10): Observable<Page<RoommateAnnouncementDTO>> {
    const headers = this.getAuthHeaders();
    const params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString());
    
    return this.http.get<Page<RoommateAnnouncementDTO>>(`${this.apiUrl}/announcements/my`, { headers, params })
      .pipe(
        tap(result => console.log(`Loaded ${result.content.length} of my announcements`)),
        catchError(this.handleError)
      );
  }

  /**
   * Get announcement by ID
   * GET /api/v1/roommates/announcements/{id}
   */
  getAnnouncementById(announcementId: number): Observable<RoommateAnnouncementDTO> {
    const headers = this.getAuthHeaders();
    return this.http.get<RoommateAnnouncementDTO>(`${this.apiUrl}/announcements/${announcementId}`, { headers })
      .pipe(
        tap(result => console.log('Loaded announcement:', result)),
        catchError(this.handleError)
      );
  }

  // ========== APPLICATION OPERATIONS ==========

  /**
   * Apply to roommate announcement with ML compatibility scoring
   * POST /api/v1/roommates/applications
   */
  applyToAnnouncement(application: RoommateApplicationCreateDTO): Observable<RoommateApplicationDTO> {
    const headers = this.getAuthHeaders();
    return this.http.post<RoommateApplicationDTO>(`${this.apiUrl}/applications`, application, { headers })
      .pipe(
        tap(result => console.log('Submitted roommate application with compatibility score:', result.compatibilityScore)),
        catchError(this.handleError)
      );
  }

  /**
   * Get applications for a specific announcement (for announcement poster)
   * GET /api/v1/roommates/applications/received
   */
  getApplicationsForAnnouncement(announcementId: number, page = 0, size = 10): Observable<Page<RoommateApplicationDTO>> {
    const headers = this.getAuthHeaders();
    const params = new HttpParams()
      .set('announcementId', announcementId.toString())
      .set('page', page.toString())
      .set('size', size.toString());
    
    return this.http.get<Page<RoommateApplicationDTO>>(`${this.apiUrl}/applications/received`, { headers, params })
      .pipe(
        tap(result => console.log(`Loaded ${result.content.length} applications for announcement`)),
        catchError(this.handleError)
      );
  }

  /**
   * Get my applications (applications I submitted)
   * GET /api/v1/roommates/applications/sent
   */
  getMyApplications(page = 0, size = 10): Observable<Page<RoommateApplicationDTO>> {
    const headers = this.getAuthHeaders();
    const params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString());
    
    return this.http.get<Page<RoommateApplicationDTO>>(`${this.apiUrl}/applications/sent`, { headers, params })
      .pipe(
        tap(result => console.log(`Loaded ${result.content.length} of my applications`)),
        catchError(this.handleError)
      );
  }

  /**
   * Respond to roommate application (accept/reject)
   * PUT /api/v1/roommates/applications/{applicationId}/respond
   */
  respondToApplication(applicationId: number, response: RoommateApplicationResponseDTO): Observable<RoommateApplicationDTO> {
    const headers = this.getAuthHeaders();
    return this.http.put<RoommateApplicationDTO>(`${this.apiUrl}/applications/${applicationId}/respond`, response, { headers })
      .pipe(
        tap(result => console.log(`Application ${response.status.toLowerCase()}:`, result)),
        catchError(this.handleError)
      );
  }

  /**
   * Get closed deals for student (from existing inquiry system)
   * GET /api/v1/roommates/closed-deals
   */
  getClosedDealsForStudent(page = 0, size = 10): Observable<Page<any>> {
    const headers = this.getAuthHeaders();
    const params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString());
    
    return this.http.get<Page<any>>(`${this.apiUrl}/closed-deals`, { headers, params })
      .pipe(
        tap(result => console.log(`Loaded ${result.content.length} closed deals`)),
        catchError(this.handleError)
      );
  }

  // ========== UTILITY METHODS ==========

  /**
   * Get compatibility level description from score
   */
  getCompatibilityLevel(score: number): string {
    if (score >= 0.90) return 'Excellent';
    if (score >= 0.75) return 'Very Good';
    if (score >= 0.60) return 'Good';
    if (score >= 0.40) return 'Fair';
    return 'Poor';
  }

  /**
   * Get compatibility percentage for display
   */
  getCompatibilityPercentage(score: number): number {
    return Math.round(score * 100);
  }

  /**
   * Get CSS class for compatibility level
   */
  getCompatibilityClass(score: number): string {
    if (score >= 0.90) return 'compatibility-excellent';
    if (score >= 0.75) return 'compatibility-very-good';
    if (score >= 0.60) return 'compatibility-good';
    if (score >= 0.40) return 'compatibility-fair';
    return 'compatibility-poor';
  }

  /**
   * Error handler
   */
  private handleError(error: any) {
    console.error('RoommateService error:', error);
    
    let errorMessage = 'An error occurred';
    
    if (error.status === 401) {
      errorMessage = 'You must be logged in to perform this action';
    } else if (error.status === 403) {
      errorMessage = 'You do not have permission to perform this action';
    } else if (error.status === 404) {
      errorMessage = 'The requested resource was not found';
    } else if (error.status === 409) {
      errorMessage = 'This action conflicts with existing data';
    } else if (error.error?.message) {
      errorMessage = error.error.message;
    } else if (error.message) {
      errorMessage = error.message;
    }
    
    return throwError(() => errorMessage);
  }
} 